lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,559 @@
+  int GPIdx = 0;
+
+  /// Convert the protocol arguments of F to generic ones.
+  for (auto const &IdxIt : ExistentialArgDescriptor) {
+    int Idx = IdxIt.first;
+    auto &param = params[Idx];
+    auto PType = param.getType();
+    assert(PType.isExistentialType());
+    /// Generate new generic parameter.
+    auto *NewGenericParam = GenericTypeParamType::get(Depth, GPIdx++, Ctx);
+    Builder.addGenericParameter(NewGenericParam);
+    Requirement NewRequirement(RequirementKind::Conformance, NewGenericParam,
+                               PType);
+    auto Source =
+        GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+    Builder.addRequirement(NewRequirement, Source, Mod);
+    ArgToGenericTypeMap.insert(
+        std::pair<int, GenericTypeParamType *>(Idx, NewGenericParam));
+    assert(ArgToGenericTypeMap.find(Idx) != ArgToGenericTypeMap.end());
+  }
+}
+
+/// Create the signature for the newly generated protocol constrained generic
+/// function.
+CanSILFunctionType
+ExistentialTransform::createExistentialSpecializedFunctionType() {
+  auto FTy = F->getLoweredFunctionType();
+  SILModule &M = F->getModule();
+  auto &Ctx = M.getASTContext();
+  GenericSignature *NewGenericSig;
+  GenericEnvironment *NewGenericEnv;
+
+  // Form a new generic signature based on the old one.
+  GenericSignatureBuilder Builder(Ctx);
+
+  /// If the original function is generic, then maintain the same.
+  auto OrigGenericSig = FTy->getGenericSignature();
+  /// First, add the old generic signature.
+  Builder.addGenericSignature(OrigGenericSig);
+
+  /// Convert existential argument types to generic argument types.
+  convertExistentialArgTypesToGenericArgTypes(Builder);
+
+  /// Compute the updated generic signature.
+  NewGenericSig = std::move(Builder).computeGenericSignature(
+      SourceLoc(), /*allowConcreteGenericParams=*/true);
+  NewGenericEnv = NewGenericSig->createGenericEnvironment();
+
+  /// Create a lambda for GenericParams.
+  auto getCanonicalType = [&](Type t) -> CanType {
+    return t->getCanonicalType(NewGenericSig);
+  };
+
+  /// Original list of parameters
+  SmallVector<SILParameterInfo, 4> params;
+  params.append(FTy->getParameters().begin(), FTy->getParameters().end());
+
+  /// Create the complete list of parameters.
+  int Idx = 0;
+  llvm::SmallVector<SILParameterInfo, 8> InterfaceParams;
+  InterfaceParams.reserve(params.size());
+  for (auto &param : params) {
+    auto iter = ArgToGenericTypeMap.find(Idx);
+    if (iter != ArgToGenericTypeMap.end()) {
+      auto GenericParam = iter->second;
+      InterfaceParams.push_back(SILParameterInfo(getCanonicalType(GenericParam),
+                                                 param.getConvention()));
+    } else {
+      InterfaceParams.push_back(param);
+    }
+    Idx++;
+  }
+
+  /// Now add the errors.
+  Optional<SILResultInfo> InterfaceErrorResult;
+  if (FTy->hasErrorResult()) {
+    InterfaceErrorResult = FTy->getErrorResult();
```

Comment:
In a follow-up PR, please remove this whole block. Instead, just pass FTy->getErrorResult() to SILFunctionType::get() below.